{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from warnings import simplefilter\n",
    "simplefilter(action=\"ignore\", category=FutureWarning)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_table(\"potential_mbps.tsv\")\n",
    "no_evid_seqs = set(df['seq_id'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "records = []\n",
    "species = ['human', 'worm', 'yeast', 'zebrafish']\n",
    "for s in species:\n",
    "    df = pd.read_table(f\"../metal_type/{s}_pred_mbps_with_type.tsv\")\n",
    "    for seq_id, df_seq in df.groupby(['seq_id']):\n",
    "        if seq_id not in no_evid_seqs: continue\n",
    "        \n",
    "        metal_positions = sorted(df_seq['resi_seq_posi'])\n",
    "        metals_str = set(df_seq['pred'])\n",
    "        metals = set()\n",
    "        for i in metals_str:\n",
    "            for j in i.split(\";\"):\n",
    "                metals.add(j)\n",
    "        \n",
    "        records.append({\n",
    "            \"seq_id\": seq_id,\n",
    "            \"species\": s,\n",
    "            \"pred\": \";\".join(metals),\n",
    "            \"resi_seq_posi\": \";\".join([str(i) for i in metal_positions])\n",
    "        })\n",
    "df = pd.DataFrame(records)\n",
    "del records"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "from itertools import combinations, product\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio import SwissProt\n",
    "from Bio.Seq import Seq\n",
    "from pathlib import Path\n",
    "\n",
    "swiss_file_path = Path(\"~/database/uniprot/swiss/\").expanduser()   \n",
    "def get_go_term(\n",
    "    df: pd.DataFrame,\n",
    ") -> pd.DataFrame:\n",
    "    \n",
    "    def gen_swiss_record(source):\n",
    "        swiss_records = SwissProt.parse(source)\n",
    "        for swiss_record in swiss_records:\n",
    "            # Convert the SwissProt record to a SeqRecord\n",
    "            record = SeqRecord(\n",
    "                Seq(swiss_record.sequence),\n",
    "                id=swiss_record.accessions[0],\n",
    "                name=swiss_record.entry_name,\n",
    "                description=str(len(swiss_record.references)), # store the number of papers involved\n",
    "                features=swiss_record.features,\n",
    "            )\n",
    "            for cross_reference in swiss_record.cross_references:\n",
    "                if len(cross_reference) < 2:\n",
    "                    continue\n",
    "                database, accession = cross_reference[:2]\n",
    "                description = cross_reference[2] if len(cross_reference) >= 3 else \"\"\n",
    "\n",
    "                dbxref = f\"{database}; {accession}; {description}\"\n",
    "                if dbxref not in record.dbxrefs:\n",
    "                    record.dbxrefs.append(dbxref)\n",
    "            yield record\n",
    "    \n",
    "    species = set(df['species'])\n",
    "    uniprots = set(df['seq_id'])\n",
    "    \n",
    "    records = []\n",
    "    for s in species:\n",
    "        swiss_file = swiss_file_path / f\"{s}.txt\"\n",
    "        for r in gen_swiss_record(swiss_file):\n",
    "            if r.id in uniprots:\n",
    "                seq_len = len(r.seq)\n",
    "                go_terms = []\n",
    "                \n",
    "                for dbxref in r.dbxrefs:\n",
    "                    database, id, desc = dbxref.split(\"; \")\n",
    "                    if database == \"GO\":\n",
    "                        go_terms.append(desc)\n",
    "                    \n",
    "                records.append({\n",
    "                    \"seq_id\": r.id,\n",
    "                    \"seq\": str(r.seq),\n",
    "                    \"length\": seq_len,\n",
    "                    \"go\": \";\".join(go_terms),\n",
    "                    \"name\": r.name,\n",
    "                    \"cnt_citation\": int(r.description),\n",
    "                })\n",
    "                \n",
    "    df_go = pd.DataFrame(records)\n",
    "    \n",
    "    return pd.merge(df, df_go, how='outer')\n",
    "\n",
    "\n",
    "from typing import List\n",
    "from Bio.PDB.PDBParser import PDBParser\n",
    "from Bio.PDB.Residue import Residue\n",
    "from Bio.SeqUtils import seq1\n",
    "import networkx as nx\n",
    "\n",
    "def has_suitable_site(\n",
    "    pdb_file: str,\n",
    "    positions: List[int],\n",
    "    coordinate_radius: float = 5.\n",
    "):\n",
    "    \n",
    "    def find_ring(g: nx.Graph) -> nx.Graph:\n",
    "        ''' Remove leaf nodes from a graph '''\n",
    "        g_ = g.copy()\n",
    "        while True:\n",
    "            leaf_nodes = [n for n, degree in g_.degree() if degree < 2]\n",
    "            if len(leaf_nodes) == 0:\n",
    "                break\n",
    "            else:\n",
    "                g_.remove_nodes_from(leaf_nodes)\n",
    "        return g_\n",
    "    \n",
    "    def get_hetero_atoms(resi: Residue):\n",
    "        res_one_letter = seq1(resi.get_resname())\n",
    "        if res_one_letter == \"C\":\n",
    "            return [resi[\"SG\"]]\n",
    "        elif res_one_letter == \"H\":\n",
    "            return [resi[\"ND1\"], resi[\"NE2\"]]\n",
    "        elif res_one_letter == \"E\":\n",
    "            return [resi[\"OE1\"], resi[\"OE2\"]]\n",
    "        else:\n",
    "            return [resi[\"OD1\"], resi[\"OD2\"]]\n",
    "    \n",
    "    def calc_distance_of_hetero_atoms(resi1: Residue, resi2: Residue):\n",
    "        try:\n",
    "            atoms1 = get_hetero_atoms(resi1)\n",
    "            atoms2 = get_hetero_atoms(resi2)\n",
    "            min_distance = 100000 # big snumber\n",
    "            for i, j in product(atoms1, atoms2):\n",
    "                d = i - j\n",
    "                if d < min_distance:\n",
    "                    min_distance = d\n",
    "            return min_distance\n",
    "        except:\n",
    "            return 100000\n",
    "    \n",
    "    try:\n",
    "        residues = list(PDBParser(QUIET=True).get_structure(\"\", pdb_file).get_residues())\n",
    "    except:\n",
    "        return\n",
    "    \n",
    "    target_residues = []\n",
    "    has_non_aligned_seq = False\n",
    "    for p in positions:\n",
    "        r: Residue = residues[p]\n",
    "        if seq1(r.get_resname()) in {\"C\", \"H\", \"E\", \"D\"}:\n",
    "            target_residues.append(r)\n",
    "        else:\n",
    "            has_non_aligned_seq = True\n",
    "            break\n",
    "    if has_non_aligned_seq: return False\n",
    "    \n",
    "    g = nx.Graph()\n",
    "    for i, j in combinations(target_residues, 2):\n",
    "        d = calc_distance_of_hetero_atoms(i, j)\n",
    "        if d < 2 * coordinate_radius:\n",
    "            g.add_edge(i.get_full_id(), j.get_full_id())\n",
    "    \n",
    "    has_adjacent_residues = False\n",
    "    g = find_ring(g)\n",
    "    if len(g.nodes) >= 3:\n",
    "        has_adjacent_residues = True\n",
    "    return has_adjacent_residues\n",
    "\n",
    "from Bio.SeqUtils.ProtParam import ProteinAnalysis\n",
    "def calc_protein_param(seq: str):\n",
    "    \n",
    "    pa = ProteinAnalysis(seq)\n",
    "    return pa.gravy()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### may be important proteins with high confidence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_id</th>\n",
       "      <th>species</th>\n",
       "      <th>pred</th>\n",
       "      <th>resi_seq_posi</th>\n",
       "      <th>length</th>\n",
       "      <th>go</th>\n",
       "      <th>name</th>\n",
       "      <th>cnt_citation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>136</th>\n",
       "      <td>Q14457</td>\n",
       "      <td>human</td>\n",
       "      <td>ZN</td>\n",
       "      <td>17;20;136;139</td>\n",
       "      <td>450</td>\n",
       "      <td>C:autophagosome;C:cytoplasm;C:cytosol;C:dendri...</td>\n",
       "      <td>BECN1_HUMAN</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>299</th>\n",
       "      <td>Q8NFG4</td>\n",
       "      <td>human</td>\n",
       "      <td>ZN</td>\n",
       "      <td>10;13;81;84</td>\n",
       "      <td>579</td>\n",
       "      <td>C:centrosome;C:cilium;C:cytoplasm;C:cytosol;C:...</td>\n",
       "      <td>FLCN_HUMAN</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>P23497</td>\n",
       "      <td>human</td>\n",
       "      <td>ZN</td>\n",
       "      <td>487;489;492</td>\n",
       "      <td>879</td>\n",
       "      <td>C:chromosome, telomeric region;C:cytoplasm;C:n...</td>\n",
       "      <td>SP100_HUMAN</td>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>322</th>\n",
       "      <td>Q92597</td>\n",
       "      <td>human</td>\n",
       "      <td>ZN</td>\n",
       "      <td>62;92;94</td>\n",
       "      <td>394</td>\n",
       "      <td>C:adherens junction;C:centrosome;C:cytoplasm;C...</td>\n",
       "      <td>NDRG1_HUMAN</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>470</th>\n",
       "      <td>P14164</td>\n",
       "      <td>yeast</td>\n",
       "      <td>ZN</td>\n",
       "      <td>48;65;70;337;339</td>\n",
       "      <td>731</td>\n",
       "      <td>C:nucleotide-excision repair factor 4 complex;...</td>\n",
       "      <td>ABF1_YEAST</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     seq_id species pred     resi_seq_posi  length  \\\n",
       "136  Q14457   human   ZN     17;20;136;139     450   \n",
       "299  Q8NFG4   human   ZN       10;13;81;84     579   \n",
       "104  P23497   human   ZN       487;489;492     879   \n",
       "322  Q92597   human   ZN          62;92;94     394   \n",
       "470  P14164   yeast   ZN  48;65;70;337;339     731   \n",
       "\n",
       "                                                    go         name  \\\n",
       "136  C:autophagosome;C:cytoplasm;C:cytosol;C:dendri...  BECN1_HUMAN   \n",
       "299  C:centrosome;C:cilium;C:cytoplasm;C:cytosol;C:...   FLCN_HUMAN   \n",
       "104  C:chromosome, telomeric region;C:cytoplasm;C:n...  SP100_HUMAN   \n",
       "322  C:adherens junction;C:centrosome;C:cytoplasm;C...  NDRG1_HUMAN   \n",
       "470  C:nucleotide-excision repair factor 4 complex;...   ABF1_YEAST   \n",
       "\n",
       "     cnt_citation  \n",
       "136            54  \n",
       "299            45  \n",
       "104            41  \n",
       "322            33  \n",
       "470            31  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_exp = df[df['species'].map(lambda x: x == \"human\" or x == 'yeast')] # species: human and yeast\n",
    "df_exp = df_exp[df_exp[\"resi_seq_posi\"].map(lambda x: len(x.split(\";\")) >= 3)] # pred resi: at least 3\n",
    "\n",
    "df_exp = get_go_term(df_exp).drop(columns=['seq'])\n",
    "\n",
    "# protein name in the entry name\n",
    "df_exp = df_exp[df_exp.apply(lambda row: row['seq_id'] != row['name'].split(\"_\")[0], axis=1)]\n",
    "\n",
    "# no \"ion binding\" in go terms (protein level)\n",
    "df_exp[df_exp['go'].map(lambda x: \"ion binding\" not in x)].sort_values(by=['cnt_citation'], ascending=False).head() "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### to be validated (Zn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_id</th>\n",
       "      <th>species</th>\n",
       "      <th>pred</th>\n",
       "      <th>resi_seq_posi</th>\n",
       "      <th>length</th>\n",
       "      <th>go</th>\n",
       "      <th>name</th>\n",
       "      <th>cnt_citation</th>\n",
       "      <th>struct_suitable</th>\n",
       "      <th>gravy</th>\n",
       "      <th>stability</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>B3KNS4</td>\n",
       "      <td>human</td>\n",
       "      <td>ZN</td>\n",
       "      <td>4;6;11;13</td>\n",
       "      <td>109</td>\n",
       "      <td></td>\n",
       "      <td>B3KNS4_HUMAN</td>\n",
       "      <td>5</td>\n",
       "      <td>True</td>\n",
       "      <td>-0.529358</td>\n",
       "      <td>32.041284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>H3BPM6</td>\n",
       "      <td>human</td>\n",
       "      <td>ZN</td>\n",
       "      <td>14;15;28;31</td>\n",
       "      <td>223</td>\n",
       "      <td></td>\n",
       "      <td>MKROS_HUMAN</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>-0.526906</td>\n",
       "      <td>38.599148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>H3BU10</td>\n",
       "      <td>human</td>\n",
       "      <td>ZN</td>\n",
       "      <td>137;140;151;154</td>\n",
       "      <td>264</td>\n",
       "      <td></td>\n",
       "      <td>H3BU10_HUMAN</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "      <td>-0.150000</td>\n",
       "      <td>31.017803</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>P0DJI9</td>\n",
       "      <td>human</td>\n",
       "      <td>CA;MG</td>\n",
       "      <td>50;54;80;96</td>\n",
       "      <td>122</td>\n",
       "      <td>C:extracellular exosome;C:high-density lipopro...</td>\n",
       "      <td>SAA2_HUMAN</td>\n",
       "      <td>11</td>\n",
       "      <td>True</td>\n",
       "      <td>-0.547541</td>\n",
       "      <td>29.256557</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152</th>\n",
       "      <td>Q32NC0</td>\n",
       "      <td>human</td>\n",
       "      <td>ZN</td>\n",
       "      <td>42;45;103;106</td>\n",
       "      <td>220</td>\n",
       "      <td></td>\n",
       "      <td>CR021_HUMAN</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "      <td>-0.747273</td>\n",
       "      <td>32.486818</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349</th>\n",
       "      <td>Q96MD7</td>\n",
       "      <td>human</td>\n",
       "      <td>ZN</td>\n",
       "      <td>44;47;69;72;85;88;95;98</td>\n",
       "      <td>179</td>\n",
       "      <td></td>\n",
       "      <td>CI085_HUMAN</td>\n",
       "      <td>6</td>\n",
       "      <td>True</td>\n",
       "      <td>-0.400559</td>\n",
       "      <td>39.160335</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>365</th>\n",
       "      <td>Q99470</td>\n",
       "      <td>human</td>\n",
       "      <td>ZN;CA;MN</td>\n",
       "      <td>103;127;130</td>\n",
       "      <td>211</td>\n",
       "      <td>C:chaperone complex;C:endoplasmic reticulum;F:...</td>\n",
       "      <td>SDF2_HUMAN</td>\n",
       "      <td>4</td>\n",
       "      <td>True</td>\n",
       "      <td>-0.362085</td>\n",
       "      <td>16.892417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>399</th>\n",
       "      <td>Q9H7C9</td>\n",
       "      <td>human</td>\n",
       "      <td>ZN;MN</td>\n",
       "      <td>24;44;118;121</td>\n",
       "      <td>122</td>\n",
       "      <td>C:cytoplasm;P:positive regulation of fat cell ...</td>\n",
       "      <td>AAMDC_HUMAN</td>\n",
       "      <td>8</td>\n",
       "      <td>True</td>\n",
       "      <td>-0.395902</td>\n",
       "      <td>29.971311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>518</th>\n",
       "      <td>P39721</td>\n",
       "      <td>yeast</td>\n",
       "      <td>ZN</td>\n",
       "      <td>49;129;153;211</td>\n",
       "      <td>246</td>\n",
       "      <td>C:cytoplasm;F:hydrolase activity</td>\n",
       "      <td>AIM2_YEAST</td>\n",
       "      <td>5</td>\n",
       "      <td>True</td>\n",
       "      <td>-0.081301</td>\n",
       "      <td>37.419919</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>519</th>\n",
       "      <td>P40080</td>\n",
       "      <td>yeast</td>\n",
       "      <td>ZN</td>\n",
       "      <td>17;20;40;43</td>\n",
       "      <td>203</td>\n",
       "      <td>C:cytoplasm;C:endosome;F:ATPase activator acti...</td>\n",
       "      <td>VFA1_YEAST</td>\n",
       "      <td>5</td>\n",
       "      <td>True</td>\n",
       "      <td>-0.879310</td>\n",
       "      <td>34.606404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>581</th>\n",
       "      <td>Q06252</td>\n",
       "      <td>yeast</td>\n",
       "      <td>CA;MN</td>\n",
       "      <td>78;80;134</td>\n",
       "      <td>201</td>\n",
       "      <td>C:cytoplasm;C:nucleus;F:peptidase inhibitor ac...</td>\n",
       "      <td>YL179_YEAST</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "      <td>-0.317910</td>\n",
       "      <td>34.164677</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>585</th>\n",
       "      <td>Q06667</td>\n",
       "      <td>yeast</td>\n",
       "      <td>ZN</td>\n",
       "      <td>135;138;151</td>\n",
       "      <td>281</td>\n",
       "      <td>C:nucleus;F:ATP binding;F:inositol pentakispho...</td>\n",
       "      <td>IPK1_YEAST</td>\n",
       "      <td>5</td>\n",
       "      <td>True</td>\n",
       "      <td>-0.285053</td>\n",
       "      <td>33.579715</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     seq_id species      pred            resi_seq_posi  length  \\\n",
       "49   B3KNS4   human        ZN                4;6;11;13     109   \n",
       "57   H3BPM6   human        ZN              14;15;28;31     223   \n",
       "59   H3BU10   human        ZN          137;140;151;154     264   \n",
       "95   P0DJI9   human     CA;MG              50;54;80;96     122   \n",
       "152  Q32NC0   human        ZN            42;45;103;106     220   \n",
       "349  Q96MD7   human        ZN  44;47;69;72;85;88;95;98     179   \n",
       "365  Q99470   human  ZN;CA;MN              103;127;130     211   \n",
       "399  Q9H7C9   human     ZN;MN            24;44;118;121     122   \n",
       "518  P39721   yeast        ZN           49;129;153;211     246   \n",
       "519  P40080   yeast        ZN              17;20;40;43     203   \n",
       "581  Q06252   yeast     CA;MN                78;80;134     201   \n",
       "585  Q06667   yeast        ZN              135;138;151     281   \n",
       "\n",
       "                                                    go          name  \\\n",
       "49                                                      B3KNS4_HUMAN   \n",
       "57                                                       MKROS_HUMAN   \n",
       "59                                                      H3BU10_HUMAN   \n",
       "95   C:extracellular exosome;C:high-density lipopro...    SAA2_HUMAN   \n",
       "152                                                      CR021_HUMAN   \n",
       "349                                                      CI085_HUMAN   \n",
       "365  C:chaperone complex;C:endoplasmic reticulum;F:...    SDF2_HUMAN   \n",
       "399  C:cytoplasm;P:positive regulation of fat cell ...   AAMDC_HUMAN   \n",
       "518                   C:cytoplasm;F:hydrolase activity    AIM2_YEAST   \n",
       "519  C:cytoplasm;C:endosome;F:ATPase activator acti...    VFA1_YEAST   \n",
       "581  C:cytoplasm;C:nucleus;F:peptidase inhibitor ac...   YL179_YEAST   \n",
       "585  C:nucleus;F:ATP binding;F:inositol pentakispho...    IPK1_YEAST   \n",
       "\n",
       "     cnt_citation  struct_suitable     gravy  stability  \n",
       "49              5             True -0.529358  32.041284  \n",
       "57              1             True -0.526906  38.599148  \n",
       "59              4             True -0.150000  31.017803  \n",
       "95             11             True -0.547541  29.256557  \n",
       "152             4             True -0.747273  32.486818  \n",
       "349             6             True -0.400559  39.160335  \n",
       "365             4             True -0.362085  16.892417  \n",
       "399             8             True -0.395902  29.971311  \n",
       "518             5             True -0.081301  37.419919  \n",
       "519             5             True -0.879310  34.606404  \n",
       "581             3             True -0.317910  34.164677  \n",
       "585             5             True -0.285053  33.579715  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# species: human and yeast\n",
    "df_exp = df[df['species'].map(lambda x: x == \"human\" or x == 'yeast')]\n",
    "\n",
    "# pred resi: at least 3\n",
    "df_exp = df_exp[df_exp[\"resi_seq_posi\"].map(lambda x: len(x.split(\";\")) >= 3)]\n",
    "\n",
    "# no \"membrane\" or \"zinc\" in go terms\n",
    "df_exp = get_go_term(df_exp)\n",
    "df_exp = df_exp[df_exp[\"go\"].map(lambda x: \"membrane\" not in x and \"zinc\" not in x)]\n",
    "\n",
    "# less than 300 amino acids\n",
    "df_exp = df_exp[df_exp[\"length\"].map(lambda x: x < 300)]\n",
    "\n",
    "# has a suitable coordinated site (distance of hetero atoms no more than 6 angstrom)\n",
    "af2_pdb_file_path = Path(\"~/database/pdb/species/\").expanduser()\n",
    "df_exp['struct_suitable'] = df_exp.apply(\n",
    "    lambda row: has_suitable_site(\n",
    "            pdb_file = af2_pdb_file_path / f\"{row['species']}_af2\" / f\"AF-{row['seq_id']}-F1-model_v4.pdb\",\n",
    "            positions = [int(i) for i in row['resi_seq_posi'].split(\";\")],\n",
    "            coordinate_radius=3\n",
    "        ), axis=1\n",
    ")\n",
    "df_exp = df_exp[df_exp['struct_suitable'].apply(lambda x: x == True)]\n",
    "\n",
    "# soluble\n",
    "df_exp['gravy'] = df_exp['seq'].apply(lambda x: ProteinAnalysis(x).gravy())\n",
    "df_exp['stability'] = df_exp['seq'].apply(lambda x: ProteinAnalysis(x).instability_index())\n",
    "df_exp = df_exp[df_exp.apply(lambda row: row['gravy'] < 0 and row['stability'] < 40, axis=1)]\n",
    "\n",
    "df_exp.drop(columns=['seq']) # P40080, Q96MD7"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### to be presented (SF4 and Ca)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_id</th>\n",
       "      <th>species</th>\n",
       "      <th>pred</th>\n",
       "      <th>resi_seq_posi</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>399</th>\n",
       "      <td>Q5VUE5</td>\n",
       "      <td>human</td>\n",
       "      <td>SF4;ZN</td>\n",
       "      <td>108;114;115;119;121;122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>888</th>\n",
       "      <td>Q9UJT9</td>\n",
       "      <td>human</td>\n",
       "      <td>SF4;ZN</td>\n",
       "      <td>400;426;452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1196</th>\n",
       "      <td>H2KZ52</td>\n",
       "      <td>worm</td>\n",
       "      <td>SF4;ZN</td>\n",
       "      <td>159;165;166;170;173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2663</th>\n",
       "      <td>A0A8M1NGW3</td>\n",
       "      <td>zebrafish</td>\n",
       "      <td>SF4;ZN</td>\n",
       "      <td>18;28;39;49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3078</th>\n",
       "      <td>A0A8M3B2I2</td>\n",
       "      <td>zebrafish</td>\n",
       "      <td>SF4;ZN</td>\n",
       "      <td>102;236;248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3732</th>\n",
       "      <td>A5PL75</td>\n",
       "      <td>zebrafish</td>\n",
       "      <td>SF4;ZN</td>\n",
       "      <td>94;100;101;105;107;108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4011</th>\n",
       "      <td>Q5U3S9</td>\n",
       "      <td>zebrafish</td>\n",
       "      <td>SF4;ZN</td>\n",
       "      <td>54;57;63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4127</th>\n",
       "      <td>Q7T2F0</td>\n",
       "      <td>zebrafish</td>\n",
       "      <td>SF4</td>\n",
       "      <td>6;31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4194</th>\n",
       "      <td>Q9YHT4</td>\n",
       "      <td>zebrafish</td>\n",
       "      <td>SF4</td>\n",
       "      <td>7;31</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          seq_id    species    pred            resi_seq_posi\n",
       "399       Q5VUE5      human  SF4;ZN  108;114;115;119;121;122\n",
       "888       Q9UJT9      human  SF4;ZN              400;426;452\n",
       "1196      H2KZ52       worm  SF4;ZN      159;165;166;170;173\n",
       "2663  A0A8M1NGW3  zebrafish  SF4;ZN              18;28;39;49\n",
       "3078  A0A8M3B2I2  zebrafish  SF4;ZN              102;236;248\n",
       "3732      A5PL75  zebrafish  SF4;ZN   94;100;101;105;107;108\n",
       "4011      Q5U3S9  zebrafish  SF4;ZN                 54;57;63\n",
       "4127      Q7T2F0  zebrafish     SF4                     6;31\n",
       "4194      Q9YHT4  zebrafish     SF4                     7;31"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df['pred'].map(lambda x: \"SF4\" in x)] # Q5VUE5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_id</th>\n",
       "      <th>species</th>\n",
       "      <th>pred</th>\n",
       "      <th>resi_seq_posi</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>945</th>\n",
       "      <td>A0A0K2WPR7</td>\n",
       "      <td>worm</td>\n",
       "      <td>CA</td>\n",
       "      <td>192;194;197;239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>955</th>\n",
       "      <td>A0A0M7RF59</td>\n",
       "      <td>worm</td>\n",
       "      <td>CA</td>\n",
       "      <td>70;135;136;194</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>976</th>\n",
       "      <td>A0A4V0INN0</td>\n",
       "      <td>worm</td>\n",
       "      <td>CA</td>\n",
       "      <td>608;610;614;655</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>980</th>\n",
       "      <td>A0A5K1IB90</td>\n",
       "      <td>worm</td>\n",
       "      <td>CA</td>\n",
       "      <td>46;48;56;82;83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1002</th>\n",
       "      <td>A3FPK9</td>\n",
       "      <td>worm</td>\n",
       "      <td>CA</td>\n",
       "      <td>40;42;53;55;56;77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2276</th>\n",
       "      <td>Q9XUU2</td>\n",
       "      <td>worm</td>\n",
       "      <td>CA</td>\n",
       "      <td>352;361;364;365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2311</th>\n",
       "      <td>Q9XWI4</td>\n",
       "      <td>worm</td>\n",
       "      <td>CA</td>\n",
       "      <td>58;118;119;175;177;183;184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2313</th>\n",
       "      <td>Q9XWL1</td>\n",
       "      <td>worm</td>\n",
       "      <td>CA</td>\n",
       "      <td>45;47;50;53;74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2324</th>\n",
       "      <td>Q9XX41</td>\n",
       "      <td>worm</td>\n",
       "      <td>CA</td>\n",
       "      <td>52;59;60;120;121;174;182;183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2339</th>\n",
       "      <td>U4PBH2</td>\n",
       "      <td>worm</td>\n",
       "      <td>CA</td>\n",
       "      <td>47;49;55;82;90</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>90 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          seq_id species pred                 resi_seq_posi\n",
       "945   A0A0K2WPR7    worm   CA               192;194;197;239\n",
       "955   A0A0M7RF59    worm   CA                70;135;136;194\n",
       "976   A0A4V0INN0    worm   CA               608;610;614;655\n",
       "980   A0A5K1IB90    worm   CA                46;48;56;82;83\n",
       "1002      A3FPK9    worm   CA             40;42;53;55;56;77\n",
       "...          ...     ...  ...                           ...\n",
       "2276      Q9XUU2    worm   CA               352;361;364;365\n",
       "2311      Q9XWI4    worm   CA    58;118;119;175;177;183;184\n",
       "2313      Q9XWL1    worm   CA                45;47;50;53;74\n",
       "2324      Q9XX41    worm   CA  52;59;60;120;121;174;182;183\n",
       "2339      U4PBH2    worm   CA                47;49;55;82;90\n",
       "\n",
       "[90 rows x 4 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df.apply(lambda row: row['species'] == 'worm' and row['pred'] == \"CA\" and len(row['resi_seq_posi'].split(\";\")) >= 4, axis=1)] # Q9XWL1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_id</th>\n",
       "      <th>species</th>\n",
       "      <th>pred</th>\n",
       "      <th>resi_seq_posi</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2104</th>\n",
       "      <td>Q9BL09</td>\n",
       "      <td>worm</td>\n",
       "      <td>FE</td>\n",
       "      <td>92;111;112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2378</th>\n",
       "      <td>P25362</td>\n",
       "      <td>yeast</td>\n",
       "      <td>FE</td>\n",
       "      <td>44;82;142;169;204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2549</th>\n",
       "      <td>Q06490</td>\n",
       "      <td>yeast</td>\n",
       "      <td>FE</td>\n",
       "      <td>396;433;488;560</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4001</th>\n",
       "      <td>Q5PRC6</td>\n",
       "      <td>zebrafish</td>\n",
       "      <td>FE</td>\n",
       "      <td>260;330;523</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      seq_id    species pred      resi_seq_posi\n",
       "2104  Q9BL09       worm   FE         92;111;112\n",
       "2378  P25362      yeast   FE  44;82;142;169;204\n",
       "2549  Q06490      yeast   FE    396;433;488;560\n",
       "4001  Q5PRC6  zebrafish   FE        260;330;523"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df.apply(lambda row: row['pred'] == \"FE\" and len(row['resi_seq_posi'].split(\";\")) >= 3, axis=1)] # P25362"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_id</th>\n",
       "      <th>species</th>\n",
       "      <th>pred</th>\n",
       "      <th>resi_seq_posi</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>426</th>\n",
       "      <td>Q6NT04</td>\n",
       "      <td>human</td>\n",
       "      <td>MN</td>\n",
       "      <td>177;295;300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1260</th>\n",
       "      <td>O02232</td>\n",
       "      <td>worm</td>\n",
       "      <td>MN</td>\n",
       "      <td>137;139;204;289</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1421</th>\n",
       "      <td>O45841</td>\n",
       "      <td>worm</td>\n",
       "      <td>MN</td>\n",
       "      <td>105;107;173;256;273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1427</th>\n",
       "      <td>O45951</td>\n",
       "      <td>worm</td>\n",
       "      <td>MN</td>\n",
       "      <td>176;244;288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1481</th>\n",
       "      <td>O76444</td>\n",
       "      <td>worm</td>\n",
       "      <td>MN</td>\n",
       "      <td>146;148;214;300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1482</th>\n",
       "      <td>O76446</td>\n",
       "      <td>worm</td>\n",
       "      <td>MN</td>\n",
       "      <td>160;162;228;313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1661</th>\n",
       "      <td>Q18357</td>\n",
       "      <td>worm</td>\n",
       "      <td>MN</td>\n",
       "      <td>92;104;106;193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1756</th>\n",
       "      <td>Q20177</td>\n",
       "      <td>worm</td>\n",
       "      <td>MN</td>\n",
       "      <td>25;116;302</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2043</th>\n",
       "      <td>Q93570</td>\n",
       "      <td>worm</td>\n",
       "      <td>MN</td>\n",
       "      <td>133;135;203;288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2236</th>\n",
       "      <td>Q9U2B4</td>\n",
       "      <td>worm</td>\n",
       "      <td>MN</td>\n",
       "      <td>128;169;181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2364</th>\n",
       "      <td>P14306</td>\n",
       "      <td>yeast</td>\n",
       "      <td>MN</td>\n",
       "      <td>94;110;160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2938</th>\n",
       "      <td>A0A8M2BGE0</td>\n",
       "      <td>zebrafish</td>\n",
       "      <td>MN</td>\n",
       "      <td>67;106;121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4015</th>\n",
       "      <td>Q5XJ69</td>\n",
       "      <td>zebrafish</td>\n",
       "      <td>MN</td>\n",
       "      <td>67;106;121</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          seq_id    species pred        resi_seq_posi\n",
       "426       Q6NT04      human   MN          177;295;300\n",
       "1260      O02232       worm   MN      137;139;204;289\n",
       "1421      O45841       worm   MN  105;107;173;256;273\n",
       "1427      O45951       worm   MN          176;244;288\n",
       "1481      O76444       worm   MN      146;148;214;300\n",
       "1482      O76446       worm   MN      160;162;228;313\n",
       "1661      Q18357       worm   MN       92;104;106;193\n",
       "1756      Q20177       worm   MN           25;116;302\n",
       "2043      Q93570       worm   MN      133;135;203;288\n",
       "2236      Q9U2B4       worm   MN          128;169;181\n",
       "2364      P14306      yeast   MN           94;110;160\n",
       "2938  A0A8M2BGE0  zebrafish   MN           67;106;121\n",
       "4015      Q5XJ69  zebrafish   MN           67;106;121"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df.apply(lambda row: row['pred'] == \"MN\" and len(row['resi_seq_posi'].split(\";\")) >= 3, axis=1)] # O45841"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_id</th>\n",
       "      <th>species</th>\n",
       "      <th>pred</th>\n",
       "      <th>resi_seq_posi</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>189</th>\n",
       "      <td>P0DJI8</td>\n",
       "      <td>human</td>\n",
       "      <td>MG</td>\n",
       "      <td>54;80;96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>211</th>\n",
       "      <td>P18405</td>\n",
       "      <td>human</td>\n",
       "      <td>MG</td>\n",
       "      <td>168;201;204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>356</th>\n",
       "      <td>Q5BKX5</td>\n",
       "      <td>human</td>\n",
       "      <td>MG</td>\n",
       "      <td>217;219;233</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>383</th>\n",
       "      <td>Q5T5J6</td>\n",
       "      <td>human</td>\n",
       "      <td>MG</td>\n",
       "      <td>392;484;509</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>491</th>\n",
       "      <td>Q7L9B9</td>\n",
       "      <td>human</td>\n",
       "      <td>MG</td>\n",
       "      <td>300;447;489</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>674</th>\n",
       "      <td>Q96EK7</td>\n",
       "      <td>human</td>\n",
       "      <td>MG</td>\n",
       "      <td>43;159;178;180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>836</th>\n",
       "      <td>Q9NRR6</td>\n",
       "      <td>human</td>\n",
       "      <td>MG</td>\n",
       "      <td>423;476;555</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>918</th>\n",
       "      <td>Q9Y2V0</td>\n",
       "      <td>human</td>\n",
       "      <td>MG</td>\n",
       "      <td>165;182;195;213</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>987</th>\n",
       "      <td>A0A679L8Q3</td>\n",
       "      <td>worm</td>\n",
       "      <td>MG</td>\n",
       "      <td>255;257;294;308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1307</th>\n",
       "      <td>O16723</td>\n",
       "      <td>worm</td>\n",
       "      <td>MG</td>\n",
       "      <td>230;232;271;285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1458</th>\n",
       "      <td>O62165</td>\n",
       "      <td>worm</td>\n",
       "      <td>MG</td>\n",
       "      <td>177;179;275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1578</th>\n",
       "      <td>Q09341</td>\n",
       "      <td>worm</td>\n",
       "      <td>MG</td>\n",
       "      <td>229;231;268;282</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1607</th>\n",
       "      <td>Q10044</td>\n",
       "      <td>worm</td>\n",
       "      <td>MG</td>\n",
       "      <td>224;226;277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2027</th>\n",
       "      <td>Q93199</td>\n",
       "      <td>worm</td>\n",
       "      <td>MG</td>\n",
       "      <td>181;210;288;290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2320</th>\n",
       "      <td>Q9XX03</td>\n",
       "      <td>worm</td>\n",
       "      <td>MG</td>\n",
       "      <td>269;271;272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2382</th>\n",
       "      <td>P25616</td>\n",
       "      <td>yeast</td>\n",
       "      <td>MG</td>\n",
       "      <td>7;9;231;235</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2510</th>\n",
       "      <td>P53933</td>\n",
       "      <td>yeast</td>\n",
       "      <td>MG</td>\n",
       "      <td>280;282;283;398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2520</th>\n",
       "      <td>Q03219</td>\n",
       "      <td>yeast</td>\n",
       "      <td>MG</td>\n",
       "      <td>13;46;76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2525</th>\n",
       "      <td>Q03508</td>\n",
       "      <td>yeast</td>\n",
       "      <td>MG</td>\n",
       "      <td>50;52;201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2528</th>\n",
       "      <td>Q03796</td>\n",
       "      <td>yeast</td>\n",
       "      <td>MG</td>\n",
       "      <td>34;36;205;217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2532</th>\n",
       "      <td>Q04223</td>\n",
       "      <td>yeast</td>\n",
       "      <td>MG</td>\n",
       "      <td>26;216;221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2579</th>\n",
       "      <td>Q12096</td>\n",
       "      <td>yeast</td>\n",
       "      <td>MG</td>\n",
       "      <td>197;199;345;434</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2580</th>\n",
       "      <td>Q12104</td>\n",
       "      <td>yeast</td>\n",
       "      <td>MG</td>\n",
       "      <td>134;232;256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2618</th>\n",
       "      <td>A0A286Y9C9</td>\n",
       "      <td>zebrafish</td>\n",
       "      <td>MG</td>\n",
       "      <td>229;231;346;347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2764</th>\n",
       "      <td>A0A8M1Q7M6</td>\n",
       "      <td>zebrafish</td>\n",
       "      <td>MG</td>\n",
       "      <td>48;297;298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2775</th>\n",
       "      <td>A0A8M1QKT5</td>\n",
       "      <td>zebrafish</td>\n",
       "      <td>MG</td>\n",
       "      <td>26;30;144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2781</th>\n",
       "      <td>A0A8M1QR63</td>\n",
       "      <td>zebrafish</td>\n",
       "      <td>MG</td>\n",
       "      <td>26;30;144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2863</th>\n",
       "      <td>A0A8M2B3D1</td>\n",
       "      <td>zebrafish</td>\n",
       "      <td>MG</td>\n",
       "      <td>161;194;197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3287</th>\n",
       "      <td>A0A8M9PEZ4</td>\n",
       "      <td>zebrafish</td>\n",
       "      <td>MG</td>\n",
       "      <td>242;244;359;360</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3457</th>\n",
       "      <td>A0A8M9Q0Q2</td>\n",
       "      <td>zebrafish</td>\n",
       "      <td>MG</td>\n",
       "      <td>148;224;260</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3644</th>\n",
       "      <td>A0A8N7TD74</td>\n",
       "      <td>zebrafish</td>\n",
       "      <td>MG</td>\n",
       "      <td>206;208;323</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3688</th>\n",
       "      <td>A0JMN5</td>\n",
       "      <td>zebrafish</td>\n",
       "      <td>MG</td>\n",
       "      <td>248;250;310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3728</th>\n",
       "      <td>A4QP92</td>\n",
       "      <td>zebrafish</td>\n",
       "      <td>MG</td>\n",
       "      <td>206;208;323;324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3890</th>\n",
       "      <td>Q08BW6</td>\n",
       "      <td>zebrafish</td>\n",
       "      <td>MG</td>\n",
       "      <td>183;200;213;231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4012</th>\n",
       "      <td>Q5U3T0</td>\n",
       "      <td>zebrafish</td>\n",
       "      <td>MG</td>\n",
       "      <td>197;199;259</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4026</th>\n",
       "      <td>Q66HV5</td>\n",
       "      <td>zebrafish</td>\n",
       "      <td>MG</td>\n",
       "      <td>283;364;390</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          seq_id    species pred    resi_seq_posi\n",
       "189       P0DJI8      human   MG         54;80;96\n",
       "211       P18405      human   MG      168;201;204\n",
       "356       Q5BKX5      human   MG      217;219;233\n",
       "383       Q5T5J6      human   MG      392;484;509\n",
       "491       Q7L9B9      human   MG      300;447;489\n",
       "674       Q96EK7      human   MG   43;159;178;180\n",
       "836       Q9NRR6      human   MG      423;476;555\n",
       "918       Q9Y2V0      human   MG  165;182;195;213\n",
       "987   A0A679L8Q3       worm   MG  255;257;294;308\n",
       "1307      O16723       worm   MG  230;232;271;285\n",
       "1458      O62165       worm   MG      177;179;275\n",
       "1578      Q09341       worm   MG  229;231;268;282\n",
       "1607      Q10044       worm   MG      224;226;277\n",
       "2027      Q93199       worm   MG  181;210;288;290\n",
       "2320      Q9XX03       worm   MG      269;271;272\n",
       "2382      P25616      yeast   MG      7;9;231;235\n",
       "2510      P53933      yeast   MG  280;282;283;398\n",
       "2520      Q03219      yeast   MG         13;46;76\n",
       "2525      Q03508      yeast   MG        50;52;201\n",
       "2528      Q03796      yeast   MG    34;36;205;217\n",
       "2532      Q04223      yeast   MG       26;216;221\n",
       "2579      Q12096      yeast   MG  197;199;345;434\n",
       "2580      Q12104      yeast   MG      134;232;256\n",
       "2618  A0A286Y9C9  zebrafish   MG  229;231;346;347\n",
       "2764  A0A8M1Q7M6  zebrafish   MG       48;297;298\n",
       "2775  A0A8M1QKT5  zebrafish   MG        26;30;144\n",
       "2781  A0A8M1QR63  zebrafish   MG        26;30;144\n",
       "2863  A0A8M2B3D1  zebrafish   MG      161;194;197\n",
       "3287  A0A8M9PEZ4  zebrafish   MG  242;244;359;360\n",
       "3457  A0A8M9Q0Q2  zebrafish   MG      148;224;260\n",
       "3644  A0A8N7TD74  zebrafish   MG      206;208;323\n",
       "3688      A0JMN5  zebrafish   MG      248;250;310\n",
       "3728      A4QP92  zebrafish   MG  206;208;323;324\n",
       "3890      Q08BW6  zebrafish   MG  183;200;213;231\n",
       "4012      Q5U3T0  zebrafish   MG      197;199;259\n",
       "4026      Q66HV5  zebrafish   MG      283;364;390"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df.apply(lambda row: row['pred'] == \"MG\" and len(row['resi_seq_posi'].split(\";\")) >= 3, axis=1)] # A4QP92"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "metalnet",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
